graylog2-server/src/main/java/org/graylog2/inputs/PersistedInputsImpl.java

```diff
@@ -52,7 +55,11 @@ public PersistedInputsImpl(InputService inputService, ServerStatus serverStatus)
         for (Input io : inputService.allOfThisNode(serverStatus.getNodeId().toString())) {
             try {
                 final MessageInput input = inputService.getMessageInput(io);
-                result.add(input);
+                if (input.onlyOnePerCluster() && input.isGlobal() && !leaderElectionService.isLeader()) {
+                    LOG.info("Not starting 'onlyOnePerCluster' input <{}/{}>", input.getName(), input.getId());
+                } else {
```

Comment:
Is it correct for all usages of this class as iterator to skip an `#onlyOnePerCluster` input here and log this message? In `InputEventListener#leaderChanged()` we iterate over `persistedInputs` and do the `input.onlyOncePerCluster()` check again. Even though with this code here in place the input doesn't get returned by the iterator.

In `InputLauncher#launchAllPersisted()` we don't do the `#onlyOncePerCluster()` check.

I think the `PersistedInputImpl#iterator()` class can be used in different contexts, so I am not sure if we should skip inputs based on the `onlyOncePerCluster` setting of the input. This can lead to unexpected behavior.

Reply:
Thanks for pointing this out 👍 

I agree that this is maximally confusing. Ideally we'd refactor the whole input persistence / launching / event handling code but I wouldn't want to touch it at this point.

I created https://github.com/Graylog2/graylog2-server/pull/11723 to remove the new checks from `PersistedInputsImpl` again and move them to the places where we know we need to filter the inputs accordingly. The PR is still WIP though, as I need to test this a bit.